File: src/main/java/myLang/block/statement/ForStatementContext.java
package myLang.block.statement;

import myLang.ContextHandler;
import myLang.block.BlockContext;
import myLang.block.localVariable.LocalVariableDeclarationStatementContext;
import myLang.expression.ExpressionContext;
import myLang.response.BooleanResponse;
import myLang.response.EmptyResponse;
import myLang.response.MyLangException;
import myLang.response.Response;
import myLangParser.MyLangParser;

public class ForStatementContext implements ContextHandler<MyLangParser.ForStatementContext> {
    @Override
    public Response handler(MyLangParser.ForStatementContext ctx) throws MyLangException {
        BlockContext.createNestedStorage();
        new LocalVariableDeclarationStatementContext().handler(ctx.localVariableDeclarationStatement());
        Response response = new ExpressionContext().handler(ctx.expression());
        if(!(response instanceof BooleanResponse)){
            throw new MyLangException("Условие в цикле for должно быть типа boolean");
        }

        while((boolean)response.getResponse()){
            new StatementContext().handler(ctx.statement());
            new ForUpdateContext().handler(ctx.forUpdate());
            response = new ExpressionContext().handler(ctx.expression());
        }

        BlockContext.deleteNestedStorage();
        return new EmptyResponse();
    }
}

class ForUpdateContext implements ContextHandler<MyLangParser.ForUpdateContext>{

    @Override
    public Response handler(MyLangParser.ForUpdateContext ctx) throws MyLangException {
        for(MyLangParser.AssignmentContext ac : ctx.statementExpressionList().assignment()){
            new AssignmentContext().handler(ac);
        }
        return new EmptyResponse();
    }
}

File: src/main/kotlin/myLangKotlin/block/statement/ForStatementContext.kt
package myLangKotlin.block.statement

import myLangKotlin.ContextHandler
import myLangKotlin.expression.ExpressionContext
import myLangKotlin.response.BooleanResponse
import myLangKotlin.response.EmptyResponse
import myLangKotlin.response.MyLangException
import myLangKotlin.response.Response
import myLangKotlin.block.BlockContext
import myLangKotlin.block.localVariable.LocalVariableDeclarationStatementContext
import myLangParser.MyLangParser

class ForStatementContext : ContextHandler<MyLangParser.ForStatementContext> {
    @Throws(MyLangException::class)
    override fun handler(ctx: MyLangParser.ForStatementContext): Response<*> {
        BlockContext.createNestedStorage()
        LocalVariableDeclarationStatementContext().handler(ctx.localVariableDeclarationStatement())
        var response: Response<*> = ExpressionContext().handler(ctx.expression()) as? BooleanResponse
                ?: throw MyLangException("Условие в цикле for должно быть типа boolean")

        while (response.response as Boolean) {
            StatementContext().handler(ctx.statement())
            ForUpdateContext().handler(ctx.forUpdate())
            response = ExpressionContext().handler(ctx.expression())
        }

        BlockContext.deleteNestedStorage()
        return EmptyResponse()
    }
}

internal class ForUpdateContext : ContextHandler<MyLangParser.ForUpdateContext> {

    @Throws(MyLangException::class)
    override fun handler(ctx: MyLangParser.ForUpdateContext): Response<*> {
        for (ac in ctx.statementExpressionList().assignment()) {
            AssignmentContext().handler(ac)
        }
        return EmptyResponse()
    }
}
